### PR TITLE
feat: opt-in Triton kernels for HC{S,M,L} inference (#76)

### LINKED ISSUE #76: Proposal: fill the three empty `vortex/ops/hc*_interface.py` files with fused Triton FFT-conv kernels (refs #16)
Hi @garykbrixi @Zymrael — proposing to fill the three empty `vortex/ops/hc{l,m,s}_interface.py` scaffolds referenced in #16 with fused Triton kernels, following @garykbrixi's note in #69 that inference-performance contributions are welcome. Numbers below from profiling `evo2_7b` on one H100 80GB SXM.

Measurements ran against `vtx==1.0.8` on PyPI; current vortex main is [`cb229ae`](https://github.com/Zymrael/vortex/commit/cb229ae50581aa072972c33a81e23eb43a1330ab) and the three `hc*_interface.py` files are still empty there. I haven't re-profiled against post-#75 main yet — flagging that explicitly — but the kernel opportunity is orthogonal to PR #75's TE-optional fallback path and applies in either mode.

### Motivation

**1. Single-GPU inference OOMs at L=131k, well below `max_seqlen: 1048576`.** Root cause is `vortex/model/model.py::compute_filter` materializing a `(D=4096, state_size=16, L)` fp32 tensor:

| seq_len | filter tensor | fits on 80 GB? |
|---|---:|---|
| 65,536 | 17.2 GiB | yes (barely) |
| 131,072 | 34.4 GiB | no |
| 1,048,576 | 262 GiB | no |

Understood that 1M is pipeline-parallel deployment; this is context for why a tiled HCL kernel that avoids the `(D, L)` allocation would unlock longer single-GPU seq_len — common for variant scoring / small-lab workflows. With `BLOCK_L=4096`, filter-tile peak drops to ~1 GiB.

**2. All three hyena variants run unfused in the default config.** `use_flashfft: False` in `shc-evo2-7b-8k-2T-v2` means HCL skips FlashFFTConv too, not just HCM. Per forward at L=65k: **~80 kernel launches across the conv system** (6+ per layer × 27 conv layers) that can collapse to ~20 with fusion.

**3. The conv system is 21.5% of CUDA time** at L=65k:

| Layer kind | CUDA ms (5 runs) | % of total | Layers |
|---|---:|---:|---:|
| **hcl** | 16,392.7 | **12.7%** | 9 |
| **hcm** | 6,868.4 | **5.3%** | 9 |
| **hcs** | 4,522.4 | **3.5%** | 9 |
| attn | 5,592.9 | 4.3% | 5 |

Remaining 49% is TE's FP8 quantize/dequantize stack around Linear projections — out of scope. Full data: [report.md](https://github.com/AlphaKhaw/vortex-kernels/blob/main/results/baseline_profile/report.md), [stacked plot](https://github.com/AlphaKhaw/vortex-kernels/blob/main/results/baseline_profile/plots/stacked_by_layer_kind.png).

### Prior art

Read the `kernels` branch (last real commit `2cd0338`, Jan 2025): its scope is training-path CGCG + vendored Mamba `causal_conv1d`. The `HyenaMR(pass)` / `HyenaLI(pass)` stubs there confirm HCM and HCL were never started. Planning to work fresh on `main` but open to reconciling if you'd prefer a rebase.

### Proposed scope

**One PR** covering all three kernels (or three on request, HCS → HCM → HCL). Each is opt-in behind a per-interface config flag defaulting to `False`:

- **`hcl_interface.py`** — Triton kernel tiling over L, computing `h = residues * (log_poles * t).exp()` per tile, fused with the FFT-conv. Avoids `(D, L)` materialization. Gated by `use_triton_hcl`.
- **`hcm_interface.py`** — `hcm_fft_conv(u, k, D, …)` matching `fftconv_func`'s signature; fuses scale/multiply around cuFFT, collapsing 6 launches to 4. Dispatched in `engine.py::parallel_fir`'s `fir_length >= 128` branch. Gated by `use_triton_hcm`.
- **`hcs_interface.py`** — direct Triton depthwise conv for `hcs_filter_length: 7` (FFT round-trip is a net loss at this size). Gated by `use_triton_hcs`.

**Zero behavioral change when flags are off, zero API changes.** Acceptance: `max_diff < 5e-2`, `mean_diff < 5e-3` vs reference bf16 on `evo2_7b` at L ∈ {8192, 32768, 65536}; bit-exact fallback when flags are off.

### Prototype repo

All profiling, kernel code, correctness tests, results at https://github.com/AlphaKhaw/vortex-kernels. Monkey-patches vortex at import time so I could iterate without forking. The profiler itself (`benchmarks/profile_evo2.py`) is reusable.

### Questions

1. Fresh modules on `main`, or rebase on the `kernels` branch?
2. Per-interface flags (`use_triton_{hcl,hcm,hcs}`) or one unified `use_triton_kernels`?
3. One PR covering all three kernels, or three smaller PRs (HCS → HCM → HCL) for staged review?

Tests would follow the existing `test/test_evo2_forward.py` convention from PR #41 unless you'd prefer otherwise. No rush — holding PRs until you weigh in. Thanks for the scaffolding.

---

### Appendix — reproducibility

RunPod H100 80GB SXM. Python 3.12, torch 2.7.1 (cuda12_9 build), TE 2.3.0, flash-attn 2.8.0.post2, evo2 0.5.3, **vtx 1.0.8** (PyPI release — pre-dates PR #75). Profiler: 5 timed forwards + 3 warmup, `record_shapes=True`, `profile_memory=True`. Forward-pass run-to-run std <2% across all completed seq_lens.

### PR DESCRIPTION
**Pull request by AlphaKhaw**
Closes #76.

## Summary

Three Triton kernels for the HC{S,M,L} blocks in `vortex/model/engine.py`, gated behind three engine flags:

* `use_hcs_kernel` — depthwise short conv (HCS)
* `use_hcm_kernel` — fused FFT conv (HCM)
* `use_hcl_kernel` — fused FFT conv + filter recompute (HCL)

All three default to `False`. When enabled, evo2_7b inference runs up to 1.73× faster at long context on a single H100 (L=65536) and unlocks L=131072 on the same hardware, where the stock path OOMs.

Branched off `main`. Signature-preserving (no new required args anywhere); opt-in via `HyenaInferenceEngine` flags only.

## End-to-end results — evo2_7b forward, n=5 + 3 warmup

NVIDIA H100 80GB HBM3, CUDA 12.9, torch 2.7.1, triton 3.3.1:

| seq_len | stock fwd ms | kernels fwd ms | speedup | stock peak GB | kernels peak GB |
|---:|---:|---:|---:|---:|---:|
| 8192   | 382.2   | 288.3   | 1.33× | 18.08 | 15.34 |
| 32768  | 2065.6  | 1397.1  | 1.48× | 32.64 | 21.64 |
| 65536  | 6653.3  | 3843.4  | 1.73× | 52.04 | 32.18 |
| 131072 | **OOM** | 12701.9 | —     | OOM   | 51.13 |

HCL accounts for ~89% of the wall-clock saving; HCM ~9%; HCS <2%. HCL also unlocks L=131072: the stock `compute_filter` materializes a `(D, state_size, L)` fp32 intermediate (~34 GiB at D=4096); the tiled kernel sums the state dimension in-register and never builds that buffer.

## End-to-end correctness — evo2_7b, bf16, H100

Stock baseline vs all-kernels-on, same loaded checkpoint, fixed seed:

| seq_len | max_abs_diff | mean_abs_diff | cos_last_token | cos_sequence_mean | argmax_match_rate |
|---:|---:|---:|---:|---:|---:|
| 8192  | 3.18 | 0.0304 | 0.9999998 | 0.999798 | 99.988% |
| 32768 | 1.94 | 0.0292 | 0.9999998 | 0.999995 | 99.997% |

`cos_last_token = 0.9999998` and `argmax_match_rate ≥ 99.988%` confirm the sampling distribution is preserved. The `max_abs_diff` of 1–3 in bf16 logit units is roundoff at near-zero cells; it does not flip token decisions.

## End-to-end results — evo2_40b forward, single H200

The kernels are per-block transforms, so the 40b weights run the same dispatch path at larger D. NVIDIA H200 141GB HBM3 (SM 9.0), CUDA 12.9, torch 2.7.1, triton 3.3.1:

| seq_len | stock fwd ms | kernels fwd ms | speedup | stock peak GB | kernels peak GB |
|---:|---:|---:|---:|---:|---:|
| 8192  | 2294.1  | 1904.0  | 1.20× | 92.00  | 86.50  |
| 32768 | 13593.5 | 9106.8  | 1.49× | 121.07 | 99.06  |
| 65536 | **OOM** | 26119.0 | —     | OOM    | 120.11 |

Same unlock pattern as 7b at this scale: stock OOMs at L=65536 on the 141 GB card, while kernels hold peak to 120.11 GB. Correctness at L=8192 (stock vs all-kernels-on, bf16, seed 0): `cos_last_token = 0.9999998`, `argmax_match_rate = 98.51%`, `max_abs_diff = 2.44` — sampling distribution preserved.

## Scope and non-goals

* evo2_7b validated on a single H100 (L=8192 → L=131072) and evo2_40b on a single H200 (L=8192 → L=65536); see tables above. The kernels are agnostic to tensor/pipeline parallelism, so multi-GPU (TP/PP) sharding should benefit identically per-rank — that remains the one planned follow-up.
* No public API surface change. Flags read off `HyenaInferenceEngine` instances; `from_pretrained` / `forward` signatures are untouched.

## Test plan

13 in-repo tests under `test/test_hc{s,m,l}_{kernel,adapter,engine,e2e}.py` covering kernel correctness → adapter parity → dispatch → end-to-end logit agreement. The L=131k unlock test (`test_hcl_e2e.py::test_hcl_unlocks_131k`) skips cleanly on GPUs without enough HBM.

Reproducible benchmark + correctness sweep in the companion harness: https://github.com/AlphaKhaw/vortex-kernels. Every JSON under `results/` carries a `run_meta` envelope pinning the GPU, driver, CUDA, torch, triton, and SHAs at measurement time.

## Reproducibility

Stack used for the tables above: torch 2.7.1, triton 3.3.1, CUDA 12.9, driver 580.126.09, H100 80GB HBM3 (SM 9.0), vortex SHA `a1dcc0f`. The evo2_40b tables are the same stack on an H200 141GB (driver 570.211.01, vortex SHA `9c54bb3`).

Commands (in the harness repo):

```
pixi run python -m benchmarks.profile_evo2 --seq-lens 8192 32768 65536 131072 --triton hcs,hcm,hcl
pixi run python -m benchmarks.correctness_evo2 --models evo2_7b --seq-lens 8192 32768 --seed 0
```

## Acknowledgements

Thanks @garykbrixi for the opportunity to contribute.

### CODE REVIEW

**Comment by garykbrixi** [comment]
Thank you for your contribution @AlphaKhaw!

I have verified the speedup and correctness on an H100 and matched your values for the 7b and the 20B generation test looks good (93.35). For the 1b there is a minor deterioration on the generation test (68.0 instead of 71.65 on generation test).

Comments:
1. could you add a warning when the 1b checkpoint is selected with the kernels given this change in performance
2. guards currently are `flag and kernel is not None`, so there could be a silent fallback -- it could be good to have a warning if fallback happens
3. hcm_fft_conv takes the gelu/dropout_mask and ignores them, which could be an issue if used in gelu=True cases outside of the current evo 2 configs
